src/events/utilities/grade-calculator/grade-calculator.ts

```diff
@@ -13,16 +14,21 @@ export class GradeCalculator {
     return Math.round(percentage * 10) / 10;
   }
 
-  public static computeGradeStream(distanceStream: number[], altitudeStream: number[]): number[] {
+  public static computeGradeStream(distanceStream: (number|null)[], altitudeStream: (number|null)[]): number[] {
 
     let gradeStream = [];
 
     for (let i = 0; i < distanceStream.length; i++) {
 
-      const previousDistance = (distanceStream[i - 1]) ? distanceStream[i - 1] : 0;
-      const currentDistance = distanceStream[i];
-      const previousAltitude = (altitudeStream[i - 1]) ? altitudeStream[i - 1] : altitudeStream[i];
-      const currentAltitude = altitudeStream[i];
+      const previousDistance = distanceStream[i - 1] || 0;
+      const currentDistance = distanceStream[i] || previousDistance; // no isNumber check needed
+      let previousAltitude = isNumber(altitudeStream[i - 1]) ? altitudeStream[i - 1] : altitudeStream[i];
+      if (!isNumber(previousAltitude)) {
```

Comment:
Let grade be 0 if there is no previous altitude then you assume that the current was used